sw/airborne/firmwares/rotorcraft/guidance/guidance_hybrid.c

```diff
@@ -0,0 +1,397 @@
+#define INT32_ANGLE_HIGH_RES_FRAC 18
+
+// Variables used for settings
+int32_t guidance_hybrid_norm_ref_airspeed;
+float alt_pitch_gain = 0.3;
+int32_t max_airspeed = MAX_AIRSPEED;
+int32_t wind_gain;
+int32_t horizontal_speed_gain;
+float max_turn_bank;
+float turn_bank_gain;
+
+// Private variables
+static struct Int32Eulers guidance_hybrid_ypr_sp;
+static struct Int32Vect2 guidance_hybrid_airspeed_sp;
+static struct Int32Vect2 guidance_h_pos_err;
+static struct Int32Vect2 guidance_hybrid_airspeed_ref;
+static struct Int32Vect2 wind_estimate;
+static struct Int32Vect2 wind_estimate_high_res;
+static struct Int32Vect2 guidance_hybrid_ref_airspeed;
+
+static int32_t norm_sp_airspeed_disp;
+static int32_t heading_diff_disp;
+static int32_t omega_disp;
+static int32_t high_res_psi;
+static int32_t airspeed_sp_heading_disp;
+static bool guidance_hovering;
+static bool force_forward_flight;
+static int32_t v_control_pitch;
+
+#if PERIODIC_TELEMETRY
+#include "subsystems/datalink/telemetry.h"
+
+static void send_hybrid_guidance(struct transport_tx *trans, struct link_device *dev) {
+  struct NedCoor_i* pos = stateGetPositionNed_i();
+  struct NedCoor_i* speed = stateGetSpeedNed_i();
+  pprz_msg_send_HYBRID_GUIDANCE(trans, dev, AC_ID,
+  &(pos->x), &(pos->y),
+  &(speed->x), &(speed->y),
+  &wind_estimate.x, &wind_estimate.y,
+  &guidance_h_pos_err.x,
+  &guidance_h_pos_err.y,
+  &guidance_hybrid_airspeed_sp.x,
+  &guidance_hybrid_airspeed_sp.y,
+  &guidance_hybrid_norm_ref_airspeed,
+  &heading_diff_disp,
+  &guidance_hybrid_ypr_sp.phi,
+  &guidance_hybrid_ypr_sp.theta,
+  &guidance_hybrid_ypr_sp.psi);
+}
+
+#endif
+
+void guidance_hybrid_init(void) {
+
+  INT_EULERS_ZERO(guidance_hybrid_ypr_sp);
+  INT_VECT2_ZERO(guidance_hybrid_airspeed_sp);
+  INT_VECT2_ZERO(guidance_hybrid_airspeed_ref);
+
+  high_res_psi = 0;
+  guidance_hovering = true;
+  horizontal_speed_gain = 8;
+  guidance_hybrid_norm_ref_airspeed = 0;
+  max_turn_bank = 40.0;
+  turn_bank_gain = 0.8;
+  wind_gain = 64;
+  force_forward_flight = 0;
+  INT_VECT2_ZERO(wind_estimate);
+  INT_VECT2_ZERO(guidance_hybrid_ref_airspeed);
+  INT_VECT2_ZERO(wind_estimate_high_res);
+
+#if PERIODIC_TELEMETRY
+  register_periodic_telemetry(DefaultPeriodic, PPRZ_MSG_ID_HYBRID_GUIDANCE, send_hybrid_guidance);
+#endif
+
+}
+
+#define INT32_ANGLE_HIGH_RES_NORMALIZE(_a) {             \
+  while ((_a) > (INT32_ANGLE_PI << (INT32_ANGLE_HIGH_RES_FRAC - INT32_ANGLE_FRAC)))  (_a) -= (INT32_ANGLE_2_PI << (INT32_ANGLE_HIGH_RES_FRAC - INT32_ANGLE_FRAC));    \
+  while ((_a) < (-INT32_ANGLE_PI << (INT32_ANGLE_HIGH_RES_FRAC - INT32_ANGLE_FRAC))) (_a) += (INT32_ANGLE_2_PI << (INT32_ANGLE_HIGH_RES_FRAC - INT32_ANGLE_FRAC));    \
+}
+
+void guidance_hybrid_run(void) {
+  guidance_hybrid_determine_wind_estimate();
+  guidance_hybrid_position_to_airspeed();
+  guidance_hybrid_airspeed_to_attitude(&guidance_hybrid_ypr_sp);
+  guidance_hybrid_set_cmd_i(&guidance_hybrid_ypr_sp);
+
+  if ( guidance_h_get_pos_err() != NULL )
```

Comment:
Which memcpy are you referring to? This is from the very old branch I'm trying to merge, so if you can give me a line number I'll use assignments instead no worries :+1: